File: client/ios/Data/Notification/NotificationHttpGateway.swift
import Foundation

public struct NotificationHttpGateway: NotificationGateway {
    private let baseUrl: String
    private let session: URLSession

    public init(baseUrl: String, session: URLSession = .shared) {
        self.baseUrl = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.session = session
    }

    public func inbox() async throws -> NotificationInboxPayload {
        try await request(path: "/notification", method: "GET", body: Optional<EmptyBody>.none)
    }

    public func unreadCount() async throws -> Int {
        let payload: UnreadPayload = try await request(path: "/notification/unread/count", method: "GET", body: Optional<EmptyBody>.none)
        return payload.unreadCount
    }

    public func markRead(ids: [String]) async throws -> Int {
        let payload: UnreadPayload = try await request(path: "/notification/mark/read", method: "POST", body: MarkReadBody(ids: ids))
        return payload.unreadCount
    }

    public func subscription(request payload: NotificationSubscriptionRequest) async throws -> Bool {
        let response: SubscriptionPayload = try await request(path: "/notification/subscription", method: "POST", body: payload)
        return response.ok
    }

    private func request<Response: Decodable, Body: Encodable>(path: String, method: String, body: Body?) async throws -> Response {
        guard let url = URL(string: baseUrl + path) else { throw NotificationGatewayError.invalidBaseUrl }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONEncoder().encode(body)
        }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw NotificationGatewayError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let error = try? JSONDecoder().decode(ErrorPayload.self, from: data)
            throw NotificationGatewayError.requestFailed(error?.message ?? "Notification request failed with HTTP \(http.statusCode).")
        }
        return try JSONDecoder().decode(Response.self, from: data)
    }
}

private struct UnreadPayload: Decodable { let unreadCount: Int }
private struct SubscriptionPayload: Decodable { let ok: Bool }
private struct MarkReadBody: Encodable { let ids: [String] }
private struct EmptyBody: Encodable {}
private struct ErrorPayload: Decodable { let message: String? }

private enum NotificationGatewayError: LocalizedError {
    case invalidBaseUrl
    case invalidResponse
    case requestFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidBaseUrl: return "Notification gateway URL is invalid."
        case .invalidResponse: return "Notification gateway returned an invalid response."
        case .requestFailed(let message): return message
        }
    }
}

File: client/android/Data/Notification/NotificationHttpGateway.kt
package app.mobiling.client.data.notification

import app.mobiling.client.contract.notification.NotificationInboxItem
import app.mobiling.client.contract.notification.NotificationInboxPayload
import app.mobiling.client.contract.notification.NotificationSubscriptionRequest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

class NotificationHttpGateway(
    private val baseUrl: String,
    private val client: OkHttpClient = OkHttpClient(),
) : NotificationGateway {
    private val jsonMediaType = "application/json".toMediaType()

    override suspend fun inbox(): NotificationInboxPayload = withContext(Dispatchers.IO) {
        val json = request("GET", "/notification", null)
        val items = json.optJSONArray("items")
        val values = buildList {
            if (items != null) {
                for (index in 0 until items.length()) {
                    items.optJSONObject(index)?.let { item ->
                        add(NotificationInboxItem(
                            id = item.optString("id"),
                            notificationId = item.optString("notificationId"),
                            status = item.optString("status", "new"),
                            title = item.optString("title", "Notification"),
                            body = item.optString("body"),
                            priority = item.optString("priority", "normal"),
                            actionUrl = item.optString("actionUrl").takeIf(String::isNotBlank),
                            createdAt = item.optString("createdAt"),
                            readAt = item.optString("readAt").takeIf(String::isNotBlank),
                        ))
                    }
                }
            }
        }
        NotificationInboxPayload(values, json.optInt("unreadCount", 0))
    }

    override suspend fun unreadCount(): Int = withContext(Dispatchers.IO) {
        request("GET", "/notification/unread/count", null).optInt("unreadCount", 0)
    }

    override suspend fun markRead(ids: List<String>): Int = withContext(Dispatchers.IO) {
        val body = JSONObject().put("ids", org.json.JSONArray(ids))
        request("POST", "/notification/mark/read", body).optInt("unreadCount", 0)
    }

    override suspend fun subscription(request: NotificationSubscriptionRequest): Boolean = withContext(Dispatchers.IO) {
        val body = JSONObject()
            .put("token", request.token)
            .put("platform", request.platform)
            .put("appKey", request.appKey)
            .put("deviceId", request.deviceId)
            .put("enabled", request.enabled)
        request("POST", "/notification/subscription", body).optBoolean("ok", false)
    }

    private fun request(method: String, path: String, body: JSONObject?): JSONObject {
        val builder = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .header("Accept", "application/json")
        if (method == "POST") builder.post((body ?: JSONObject()).toString().toRequestBody(jsonMediaType)) else builder.get()
        client.newCall(builder.build()).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val message = runCatching { JSONObject(responseBody).optString("message") }.getOrNull().orEmpty()
                error(message.ifBlank { "Notification request failed with HTTP ${response.code}." })
            }
            return JSONObject(responseBody.ifBlank { "{}" })
        }
    }
}
